tools/payloads/ysoserial/findYsoserialOffsets.rb

```diff
@@ -0,0 +1,179 @@
+  end
+  return payloadArray
+end
+
+def isLengthOffset?(currByte,nextByte)
+  # If this byte has been changed, and is different by one, then it must be a length value
+  if nextByte and currByte.position == nextByte.position and currByte.action == "-"
+    if nextByte.element.ord - currByte.element.ord == 1
+      #STDERR.puts "Found a length offset at #{currByte.position}"
+      return true
+    end
+  end
+  return false
+end
+
+def isBufferOffset?(currByte,nextByte)
+  # If this byte has been inserted, then it must be part of the increasingly large payload buffer
+  if (currByte.action == "+" and (nextByte.nil? or (currByte.position != nextByte.position)))
+    #STDERR.puts "Found a buffer offset at #{currByte.position}"
+    return true
+  end
+  return false
+end
+
+def diff(a,b)
+  diffs = []
+  obj=  Diff::LCS.diff(a,b)
+  obj.each do |i|
+    i.each do |j|
+      diffs.push(j)
+    end
+  end
+  return diffs
+end
+
+def getPayloadList()
+  # Call ysoserial and return the list of payloads that can be generated
+  payloads = `java -jar ysoserial-original.jar 2>&1`
+  payloads.encode!('ASCII', 'binary', invalid: :replace, undef: :replace, replace: '')
+  payloads = payloads.split("\n")
+
+  # Make sure the headers are intact, then skip over them
+  abort unless payloads[0] == "Y SO SERIAL?"
+  payloads = payloads.drop(5)
+
+  payloadList = []
+  # Skip the header rows
+  payloads.each do |line|
+    next unless line.include?"     "
+    payloadList.push(line.scan(/^     ([^ ]*) .*/).first.last)
+  end
+  return payloadList
+end
+
+results = {}
+payloadList = getPayloadList
+payloadList.each do |payload|
+  STDERR.puts "Generating payloads for #{payload}..."
+
+  emptyPayload = generatePayload(payload,0)
+
+  if emptyPayload.nil?
+    STDERR.puts "  ERROR: Errored while generating '#{payload}' and it will not be supported"
+    results[payload]={"status": "unsupported"}
+    next
+  end    
+
+  payloadArray = generatePayloadArray(payload)
+
+  lengthOffset = []
+  bufferOffset = []
+
+  # Comparing diffs of various payload lengths to find length and buffer offsets
+  (1..payloadArray.length-1-1).each do |i|
+    # Compare this binary with the next one
+    diffs = diff(payloadArray[i],payloadArray[i+1])
+
+    break if diffs.nil?
+ 
+    # Iterate through each diff, searching for offsets of the length and the payload
+    (0..diffs.length-1).each do |j|
+      currByte = diffs[j]
+      nextByte = diffs[j+1]
+      prevByte = diffs[j-1]
+
+      if j>0
+        # Skip this if we compared these two bytes on the previous iteration
+        next if prevByte.position == currByte.position
+      end
+
+      # Compare this byte and the following byte to identify length and buffer offsets
+      lengthOffset.push(currByte.position) if isLengthOffset?(currByte,nextByte)
+      bufferOffset.push(currByte.position) if isBufferOffset?(currByte,nextByte)
+    end
+  end
+
+  payloadBytes = Base64.encode64(emptyPayload).chomp
```

Comment:
You're using `Base64`, rather than the `Rex` base64 encoding.

Ruby's `Base64` is dumb. Observe the happy little line break in the middle of the string:

```ruby
2.3.0 :001 > require 'base64'
 => true 
2.3.0 :002 > Base64.encode64('AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA')
 => "QUFBQUFBQUFBQUFBQUFBQUFBQUFBQUFBQUFBQUFBQUFBQUFBQUFBQUFBQUFB\nQUFBQUFBQUFBQUFBQUFBQUFBQUFBQUFBQUFBQUFBQUFBQQ==\n" 
```

I'm pretty sure that's not what you want. Instead:

```suggestion
  payloadBytes = Base64.strict_encode64(emptyPayload)
```

```
2.3.0 :001 > require 'base64'
 => true 
2.3.0 :002 > Base64.strict_encode64('AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA')
 => "QUFBQUFBQUFBQUFBQUFBQUFBQUFBQUFBQUFBQUFBQUFBQUFBQUFBQUFBQUFBQUFBQUFBQUFBQUFBQUFBQUFBQUFBQUFBQUFBQUFBQUFBQQ==" 
2.3.0 :003 > 
```